File: app/src/main/java/org/intelehealth/app/dialog/BluetoothDeviceChooseDialog.java
package org.intelehealth.app.dialog;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.Dialog;
import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothClass;
import android.bluetooth.BluetoothDevice;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.AdapterView;
import android.widget.Button;
import android.widget.ListView;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.fragment.app.DialogFragment;

import com.smartcaredoc.app.R;
import org.intelehealth.app.activities.visitSummaryActivity.BluetoothDeviceAdapter;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Prajwal Maruti Waingankar on 04-07-2022, 18:00
 * Copyright (c) 2021 . All rights reserved.
 * Email: prajwalwaingankar@gmail.com
 * Github: prajwalmw
 */

public class BluetoothDeviceChooseDialog extends DialogFragment {

    private final String TAG = getClass().getSimpleName();

    private Context mContext;
    private onDeviceItemClickListener mListener;
    private ListView lvPairedDevices, lvFoundDevices;
    private TextView tvPairedDeviceEmpty, tvFoundDeviceEmpty, tvSearchDevice;
    private ProgressBar progressBar;
    private Button btn_hide;

    private BroadcastReceiver mBluetoothReceiver;
    private IntentFilter mBluetoothIntentFilter;
    private BluetoothAdapter mBluetoothAdapter;
    private BluetoothDeviceAdapter pairedDeviceAdapter, foundDeviceAdapter;
    private List<BluetoothDevice> pairedDeviceList, foundDeviceList;
    private boolean mSearchInited = false;// 若为true表示搜索设备按钮已按下过，数据已初始化
    private boolean mRegistered = false;// 若为true表示接收器已注册
    private boolean isHidePairedDevlist = false;//是否隐藏配对列表

    @Override
    public void onAttach(Activity activity) {
        super.onAttach(activity);
        mContext = activity;
    }


    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {

        LayoutInflater inflater = getActivity().getLayoutInflater();
        View view = inflater.inflate(R.layout.dialog_choose_bluetooth_device, null);
        initView(view);
        setListener();
        initData();
        AlertDialog.Builder builder = new AlertDialog.Builder(mContext);
        builder.setView(view).setCancelable(true).setNegativeButton(R.string.cancel, null);
        return builder.create();
    }

    private void initView(View view) {
        lvPairedDevices = (ListView) view.findViewById(R.id.lv_dialog_choose_bluetooth_device_paired_devices);
        tvPairedDeviceEmpty = (TextView) view.findViewById(R.id.tv_dialog_choose_bluetooth_device_paired_devices_empty);
        lvFoundDevices = (ListView) view.findViewById(R.id.lv_dialog_choose_bluetooth_device_found_devices);
        tvFoundDeviceEmpty = (TextView) view.findViewById(R.id.tv_dialog_choose_bluetooth_device_found_devices_empty);
        tvSearchDevice = (TextView) view.findViewById(R.id.tv_dialog_choose_bluetooth_device_search_device);
        progressBar = (ProgressBar) view.findViewById(R.id.pb_dialog_choose_bluetooth_device_progress_bar);
        btn_hide = view.findViewById(R.id.btn_hide);
    }

    private void setListener() {
        tvSearchDevice.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                tvSearchDevice.setEnabled(false);
                progressBar.setVisibility(View.VISIBLE);
                tvFoundDeviceEmpty.setVisibility(View.GONE);
                if (mSearchInited) {
                    foundDeviceList.clear();
                    foundDeviceAdapter.notifyDataSetChanged();
                } else {
                    foundDeviceList = new ArrayList<>();
                    foundDeviceAdapter = new BluetoothDeviceAdapter(mContext, foundDeviceList);
                    lvFoundDevices.setAdapter(foundDeviceAdapter);
                    mBluetoothReceiver = new BluetoothDeviceReceiver();
                    mBluetoothIntentFilter = new IntentFilter();
                    mBluetoothIntentFilter.addAction(BluetoothDevice.ACTION_FOUND);
                    mBluetoothIntentFilter.addAction(BluetoothAdapter.ACTION_DISCOVERY_FINISHED);
                    mSearchInited = true;
                }
                mContext.registerReceiver(mBluetoothReceiver, mBluetoothIntentFilter);
                mRegistered = true;
                mBluetoothAdapter.startDiscovery();
            }
        });

        lvPairedDevices.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                mBluetoothAdapter.cancelDiscovery();
                if (mRegistered) {
                    mContext.unregisterReceiver(mBluetoothReceiver);
                    mRegistered = false;
                }
                mListener.onDeviceItemClick((BluetoothDevice) parent.getAdapter().getItem(position));
                getDialog().dismiss();
            }
        });
        lvFoundDevices.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                mBluetoothAdapter.cancelDiscovery();
                if (mRegistered) {
                    mContext.unregisterReceiver(mBluetoothReceiver);
                    mRegistered = false;
                }
                mListener.onDeviceItemClick((BluetoothDevice) parent.getAdapter().getItem(position));
                getDialog().dismiss();
            }
        });
        btn_hide.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if(isHidePairedDevlist){//当前是隐藏
                    isHidePairedDevlist = false;
                    lvPairedDevices.setVisibility(View.VISIBLE);
                    btn_hide.setText("Hide_↑↑↑");
                }else{//当前是可见
                    isHidePairedDevlist = true;
                    lvPairedDevices.setVisibility(View.GONE);
                    btn_hide.setText("Show_↓↓↓");
                }
            }
        });
    }

    private void initData() {
        mBluetoothAdapter = BluetoothAdapter.getDefaultAdapter();
        pairedDeviceList = new ArrayList<>(mBluetoothAdapter.getBondedDevices());
        if (pairedDeviceList.size() == 0) {
            tvPairedDeviceEmpty.setVisibility(View.VISIBLE);
        }
        pairedDeviceAdapter = new BluetoothDeviceAdapter(mContext, pairedDeviceList);
        lvPairedDevices.setAdapter(pairedDeviceAdapter);
    }

    @Override
    public void onDismiss(DialogInterface dialog) {
        super.onDismiss(dialog);
        mBluetoothAdapter.cancelDiscovery();
        if (mRegistered) {
            mContext.unregisterReceiver(mBluetoothReceiver);
        }
    }

    public void setOnDeviceItemClickListener(onDeviceItemClickListener listener) {
        mListener = listener;
    }

    public interface onDeviceItemClickListener {
        void onDeviceItemClick(BluetoothDevice device);
    }

    private class BluetoothDeviceReceiver extends BroadcastReceiver {

        @Override
        public void onReceive(Context context, Intent intent) {
            String action = intent.getAction();
            // When discovery finds a device
            if (BluetoothDevice.ACTION_FOUND.equals(action)) {
                // Get the BluetoothDevice object from the Intent
                BluetoothDevice device = intent.getParcelableExtra(BluetoothDevice.EXTRA_DEVICE);
                int devType = device.getBluetoothClass().getMajorDeviceClass();
                if(devType != BluetoothClass.Device.Major.IMAGING){
                    return;
                }

                if (!foundDeviceList.contains(device)) {
                    foundDeviceList.add(device);
                    foundDeviceAdapter.notifyDataSetChanged();
                }
            } else if (BluetoothAdapter.ACTION_DISCOVERY_FINISHED.equals(action)) {
                mBluetoothAdapter.cancelDiscovery();
                mContext.unregisterReceiver(mBluetoothReceiver);
                mRegistered = false;
                tvSearchDevice.setEnabled(true);
                progressBar.setVisibility(View.GONE);
                if (foundDeviceList.size() == 0) {
                    tvFoundDeviceEmpty.setVisibility(View.VISIBLE);
                }
            }
        }
    }

}

File: app/src/main/java/org/intelehealth/app/activities/prescription/thermalprinter/BluetoothDeviceChooseDialog.kt
package org.intelehealth.app.activities.prescription.thermalprinter

import android.app.Activity
import android.app.AlertDialog
import android.app.Dialog
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothClass
import android.bluetooth.BluetoothDevice
import android.content.BroadcastReceiver
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.view.View
import android.view.WindowManager
import android.widget.Button
import android.widget.ListView
import android.widget.ProgressBar
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.fragment.app.DialogFragment
import org.intelehealth.app.R

class BluetoothDeviceChooseDialog : DialogFragment() {

    private val TAG = this::class.java.simpleName

    private lateinit var mContext: Context
    private var mListener: OnDeviceItemClickListener? = null
    private lateinit var lvPairedDevices: ListView
    private lateinit var lvFoundDevices: ListView
    private lateinit var tvPairedDeviceEmpty: TextView
    private lateinit var tvFoundDeviceEmpty: TextView
    private lateinit var tvSearchDevice: TextView
    private lateinit var progressBar: ProgressBar
    private lateinit var btnHide: Button

    private var mBluetoothReceiver: BroadcastReceiver? = null
    private var mBluetoothIntentFilter: IntentFilter? = null
    private lateinit var mBluetoothAdapter: BluetoothAdapter
    private lateinit var pairedDeviceAdapter: BluetoothDeviceAdapter
    private lateinit var foundDeviceAdapter: BluetoothDeviceAdapter
    private lateinit var pairedDeviceList: MutableList<BluetoothDevice>
    private lateinit var foundDeviceList: MutableList<BluetoothDevice>
    private var mSearchInited =
        false // true if the search button has been pressed and data initialized
    private var mRegistered = false // true if the receiver is registered
    private var isHidePairedDevList = false // true if the paired device list is hidden

    override fun onAttach(activity: Activity) {
        super.onAttach(activity)
        mContext = activity
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val inflater = requireActivity().layoutInflater
        val view = inflater.inflate(R.layout.dialog_choose_bluetooth_device, null)
        initView(view)
        setListener()
        initData()
        val builder = AlertDialog.Builder(mContext)
        builder.setView(view).setCancelable(true)
        val dialog = builder.create()
        dialog.window?.apply {
            setBackgroundDrawableResource(R.drawable.ui2_rounded_corners_dialog_bg)
            addFlags(WindowManager.LayoutParams.FLAG_BLUR_BEHIND)
        }

        return dialog
    }

    private fun initView(view: View) {
        lvPairedDevices = view.findViewById(R.id.lv_dialog_choose_bluetooth_device_paired_devices)
        tvPairedDeviceEmpty =
            view.findViewById(R.id.tv_dialog_choose_bluetooth_device_paired_devices_empty)
        lvFoundDevices = view.findViewById(R.id.lv_dialog_choose_bluetooth_device_found_devices)
        tvFoundDeviceEmpty =
            view.findViewById(R.id.tv_dialog_choose_bluetooth_device_found_devices_empty)
        tvSearchDevice = view.findViewById(R.id.tv_dialog_choose_bluetooth_device_search_device)
        progressBar = view.findViewById(R.id.pb_dialog_choose_bluetooth_device_progress_bar)
        btnHide = view.findViewById(R.id.btn_hide)
    }

    private fun setListener() {
        tvSearchDevice.setOnClickListener {
            tvSearchDevice.isEnabled = false
            progressBar.visibility = View.VISIBLE
            tvFoundDeviceEmpty.visibility = View.GONE
            if (mSearchInited) {
                foundDeviceList.clear()
                foundDeviceAdapter.notifyDataSetChanged()
            } else {
                foundDeviceList = mutableListOf()
                foundDeviceAdapter = BluetoothDeviceAdapter(mContext, foundDeviceList)
                lvFoundDevices.adapter = foundDeviceAdapter
                mBluetoothReceiver = BluetoothDeviceReceiver()
                mBluetoothIntentFilter = IntentFilter().apply {
                    addAction(BluetoothDevice.ACTION_FOUND)
                    addAction(BluetoothAdapter.ACTION_DISCOVERY_FINISHED)
                }
                mSearchInited = true
            }
            ContextCompat.registerReceiver(
                mContext,
                mBluetoothReceiver!!,
                mBluetoothIntentFilter!!,
                ContextCompat.RECEIVER_NOT_EXPORTED
            )
            mRegistered = true
            mBluetoothAdapter.startDiscovery()
        }

        lvPairedDevices.setOnItemClickListener { parent, _, position, _ ->
            mBluetoothAdapter.cancelDiscovery()
            if (mRegistered) {
                mContext.unregisterReceiver(mBluetoothReceiver)
                mRegistered = false
            }
            mListener?.onDeviceItemClick(parent.getItemAtPosition(position) as BluetoothDevice)
            dialog?.dismiss()
        }

        lvFoundDevices.setOnItemClickListener { parent, _, position, _ ->
            mBluetoothAdapter.cancelDiscovery()
            if (mRegistered) {
                mContext.unregisterReceiver(mBluetoothReceiver)
                mRegistered = false
            }
            mListener?.onDeviceItemClick(parent.getItemAtPosition(position) as BluetoothDevice)
            dialog?.dismiss()
        }

        btnHide.setOnClickListener {
            if (isHidePairedDevList) {
                isHidePairedDevList = false
                lvPairedDevices.visibility = View.VISIBLE
                btnHide.text = "Hide_↑↑↑"
            } else {
                isHidePairedDevList = true
                lvPairedDevices.visibility = View.GONE
                btnHide.text = "Show_↓↓↓"
            }
        }
    }

    private fun initData() {
        mBluetoothAdapter = BluetoothAdapter.getDefaultAdapter()
        pairedDeviceList = mBluetoothAdapter.bondedDevices.toMutableList()
        if (pairedDeviceList.isEmpty()) {
            tvPairedDeviceEmpty.visibility = View.VISIBLE
        }
        pairedDeviceAdapter = BluetoothDeviceAdapter(mContext, pairedDeviceList)
        lvPairedDevices.adapter = pairedDeviceAdapter
    }

    override fun onDismiss(dialog: DialogInterface) {
        super.onDismiss(dialog)
        mBluetoothAdapter.cancelDiscovery()
        if (mRegistered) {
            mContext.unregisterReceiver(mBluetoothReceiver)
        }
    }

    fun setOnDeviceItemClickListener(listener: OnDeviceItemClickListener) {
        mListener = listener
    }

    interface OnDeviceItemClickListener {
        fun onDeviceItemClick(device: BluetoothDevice)
    }

    private inner class BluetoothDeviceReceiver : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val action = intent.action
            if (BluetoothDevice.ACTION_FOUND == action) {
                val device =
                    intent.getParcelableExtra<BluetoothDevice>(BluetoothDevice.EXTRA_DEVICE)
                val devType = device?.bluetoothClass?.majorDeviceClass
                if (devType != BluetoothClass.Device.Major.IMAGING) {
                    return
                }
                if (device !in foundDeviceList) {
                    foundDeviceList.add(device)
                    foundDeviceAdapter.notifyDataSetChanged()
                }
            } else if (BluetoothAdapter.ACTION_DISCOVERY_FINISHED == action) {
                mBluetoothAdapter.cancelDiscovery()
                mContext.unregisterReceiver(mBluetoothReceiver)
                mRegistered = false
                tvSearchDevice.isEnabled = true
                progressBar.visibility = View.GONE
                if (foundDeviceList.isEmpty()) {
                    tvFoundDeviceEmpty.visibility = View.VISIBLE
                }
            }
        }
    }
    override fun onStart() {
        super.onStart()
        dialog?.window?.apply {
            val width = requireContext().resources.getDimensionPixelSize(R.dimen.internet_dialog_width)  // Set width of the dialog
            setLayout(width, WindowManager.LayoutParams.WRAP_CONTENT)  // Set width and height
        }
    }
}